theories/ring.elpi

```diff
@@ -314,25 +318,24 @@ quote.ring R _ TR _ {{ Negz lp:In }} {{ @RNegz lp:OutM1 }} Out VarMap :-
   quote.expr.opp { quote.expr.add { quote.expr.one } Out1 } Out.
 % Z constants
 quote.ring R _ _ _ In {{ @RZC lp:In }} Out _ :-
-  coq.unify-eq {{ ZInstances.Z_ringType }} R ok,
+  coq.unify-eq {{ Z_ringType }} R ok,
   ground-Z In, !,
   quote.expr.constant In Out.
 % morphisms
 quote.ring R _ TR Morph In
            {{ @RMorph lp:Q lp:R lp:NewMorphInst lp:OutM }} Out VarMap :-
-  NewMorph = (x\ {{ @GRing.RMorphism.apply
-                      lp:Q lp:R _ lp:NewMorphInst lp:x }}),
-  coq.unify-eq In (NewMorph In1) ok,
+  ring->sring R SR,
+  NewMorph = (x\ {{ @GRing.RMorphism.sort lp:SQ lp:SR lp:NewMorphInst lp:x }}),
+  coq.unify-eq In (NewMorph In1) ok, !, ring->sring Q SQ,
   !,
   % TODO: for concrete morphisms, should we unpack [NewMorph]?
   quote.ring Q { ring->field Q } TR (x\ Morph (NewMorph x)) In1 OutM Out VarMap.
 % additive functions
 quote.ring R _ TR Morph In
            {{ @RMorph' lp:U lp:R lp:NewMorphInst lp:OutM }} Out VarMap :-
-  NewMorph =
-    (x\ {{ @GRing.Additive.apply
-             lp:U (GRing.Ring.zmodType lp:R) _ lp:NewMorphInst lp:x }}),
-  coq.unify-eq In (NewMorph In1) ok,
+  ring->nmod R NR,
+  NewMorph = (x\ {{ @GRing.Additive.sort lp:NU lp:NR lp:NewMorphInst lp:x }}),
+  coq.unify-eq In (NewMorph In1) ok, !, zmod->nmod U NU,
```

Comment:
Calling `coq.unify-eq` 3 times seems unnecessary and wasteful. The cut operator should be placed only after `zmod->nmod U NU`.
Now I think things like `ring->nmod` should be syntactic and we should call `coq.unify-eq` explicitly. *I will fix*.